### PR TITLE
Prototype `buildSrc/groovy/js/*` scripts in Kotlin

### PR DESCRIPTION
**Pull request by yevhenii-nadtochii**
This draft is not going to be merged as approved. A separate PR into config repository would be created.

### CODE REVIEW

File: buildSrc/build.gradle.kts [hunk 1]
```diff
@@ -59,6 +59,7 @@ val jacksonVersion = "2.13.0"
 val googleAuthToolVersion = "2.1.2"
 val licenseReportVersion = "2.0"
 val grGitVersion = "3.1.1"
+val protobufPluginVersion = "0.8.17"
```
Comment (armiol) [inline buildSrc/build.gradle.kts:62]: Let's discuss this one. Either it needs a comment telling to sync this version (similar to `errorProneVersion` below), or we might not need this version at all — as (maybe) we always grab the most recent from Gradle plugin repository.

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/JsEnvironment.kt [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+ * Consists of three parts describing:
+ *
+ *  1. A module itself;
+ *  2. Tools and their input/output files;
+ *  3. Code generation.
+ */
+interface JsEnvironment {
+
+
+    // A module itself.
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/JsEnvironment.kt:44]: Two lines above and two lines below is too much.

Let's try using [the convention we have for Java](https://github.com/SpineEventEngine/SpineEventEngine.github.io/wiki/Java-Code-Style#comment-separators-for-sections-of-the-class), instead.

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/JsEnvironment.kt [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.internal.gradle.javascript
+
+import java.io.File
+import org.apache.tools.ant.taskdefs.condition.Os
+
+/**
+ * Information about JavaScript environment.
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/JsEnvironment.kt:33]: This is too generic. Smells like browser to me.

How about this:

`Describes the environment in which JavaScript code is assembled and processed during the build.`

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/JsEnvironment.kt [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    /**
+     * An access token that allows installation and/or publishing modules.
+     *
+     * Default value is read from the environmental variable - `NPM_TOKEN`.
+     * "PUBLISHING_FORBIDDEN" stub value would be assigned in case `NPM_TOKEN` variable is not set.
+     *
+     * During installation a token is required only if dependencies
+     * from private repositories are used.
+     *
+     * See [Creating and viewing access tokens | npm Docs](https://docs.npmjs.com/creating-and-viewing-access-tokens)
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/JsEnvironment.kt:113]: Let's end all sentences with periods.

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/JsEnvironment.kt [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+
+/**
+ * Allows overriding [JsEnvironment]'s defaults.
+ *
+ * All of declared properties can be split into two groups:
+ *
+ *  1. The ones that *define* something - can be overridden;
+ *  2. The ones that *describe* something - can not be overridden.
+ *
+ *  Overriding of "describing" properties leads to inconsistency with expectations. They are not
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/JsEnvironment.kt:198]: Please rephrase this section. Something (which probably should tell how "describing" are used, and why they are NOT used by `npm`) is missing.

Also:
```
`npm` tool
```
and 

```
JS
```

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/JsExtension.kt [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+ * javascript {
+ *     environment {
+ *         moduleVersion = "$moduleVersion-SPECIAL_EDITION"
+ *     }
+ * }
+ * ```
+ *
+ * ### Tasks and Plugins
+ *
+ * The main spirit of tasks configuration in this scope is extracting procedural code into
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/JsExtension.kt:86]: Those who read this won't know anything about "procedural" code. It just does not exist for them anymore. Let's simplify this thought accordingly.

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/plugin/Idea.kt [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.internal.gradle.javascript.plugin
+
+import org.gradle.kotlin.dsl.configure
+import org.gradle.plugins.ide.idea.model.IdeaModel
+
+/**
+ * Applies and configures an `IDEA` plugin to work with a JavaScript module.
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/plugin/Idea.kt:33]: It's either
```
IDEA plugin
```
or 
```
`idea` plugin
```

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/plugin/JsPlugins.kt [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+import io.spine.internal.gradle.javascript.JsEnvironment
+import org.gradle.api.Project
+import org.gradle.api.tasks.TaskContainer
+
+/**
+ * A scope for applying and configuring JavaScript-related plugins.
+ *
+ * The scope extends [JsContext] and provides access to the project's [TaskContainer].
+ *
+ * Supposing, one needs to apply and configure `FooBar` plugin. To achieve that,
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/plugin/JsPlugins.kt:45]:
```
Let's imagine one wants to apply and configure some `FooBar` plugin. To do that, she needs to go through several steps:

...
```

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/plugin/JsPlugins.kt [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+ * ```
+ * fun JsPlugins.fooBar() {
+ *     plugins.apply("com.fooBar")
+ *     extensions.configure<FooBarExtension> {
+ *         // ...
+ *     }
+ * }
+ * ```
+ *
+ * And how to apply it in `build.gradle.kts`:
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/plugin/JsPlugins.kt:63]: `And here's how to ...`

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/plugin/McJs.kt [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+package io.spine.internal.gradle.javascript.plugin
+
+import io.spine.internal.gradle.javascript.task.assembleJs
+import io.spine.internal.gradle.javascript.task.compileProtoToJs
+import org.gradle.api.Task
+import org.gradle.kotlin.dsl.withGroovyBuilder
+
+/**
+ * Applies and configures `MsJsPlugin`.
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/plugin/McJs.kt:35]: It's `McJsPlugin`.

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/plugin/Protobuf.kt [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+import com.google.protobuf.gradle.remove
+import io.spine.internal.dependency.Protobuf
+import io.spine.internal.gradle.javascript.task.compileProtoToJs
+
+/**
+ * Applies and configures `Protobuf` plugin to work with a JavaScript module.
+ *
+ * In particular, this method:
+ *
+ *  1. Specifies `protoc` compiler;
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/plugin/Protobuf.kt:43]: What do you mean? "Configures"? 

Let's discuss this if you need.

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/plugin/Protobuf.kt [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+        }
+
+        generateProtoTasks {
+            all().forEach { task ->
+
+                task.builtins {
+
+                    // Java is not needed in this project.
+
+                    remove("java")
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/plugin/Protobuf.kt:70]: I think this step may have unwanted consequences. At this point we don't know whether Java is required.

Let's discuss if needed.

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/task/Assemble.kt [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+    installNodePackages()
+    compileProtoToJs()
+    updatePackageVersion()
+
+    assembleJs().also {
+        assemble.configure {
+            dependsOn(it)
+        }
+    }
+
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/task/Assemble.kt:72]: Please kill.

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/task/Assemble.kt [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+        )
+    }
+
+
+/**
+ * Locates `compileProtoToJs` task in this [TaskContainer].
+ *
+ * The task compiles Protobuf messages into JavaScript.
+ *
+ * This is a lifecycle task that performs no action itself. It is used to aggregate other tasks
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/task/Assemble.kt:103]: Let's settle down. Either it "performs no action itself" or  "compiles Protobuf messages".

Right now the whole text looks a copy-paste mistake. Please adjust.

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/task/Assemble.kt [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+
+        description = "Compiles Protobuf messages into JavaScript."
+        group = jsAssembleTask
+    }
+
+
+/**
+ * Locates `installNodePackages` task in this [TaskContainer].
+ *
+ * The task installs a package and any packages that it depends on using the `npm install` command.
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/task/Assemble.kt:121]: I don't understand this part: "installs a package and any packages that it depends on". There is no certain "package" in this context. Therefore, there's nothing to speak of.

Please rephrase.

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/task/Check.kt [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+ *
+ * javascript {
+ *     tasks {
+ *         assemble()
+ *         check()
+ *     }
+ * }
+ * ```
+ *
+ *  @param configuration any additional configuration related to the module's verifying.
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/task/Check.kt:66]: Did you mean "verification"?

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/task/Check.kt [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+
+    checkJs().also {
+        check.configure {
+            dependsOn(it)
+        }
+    }
+
+    configuration()
+}
+
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/task/Check.kt:82]: Let's not have so many empty lines. There are of no use.

File: buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/task/Check.kt [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+        description = "Runs tests, audits NPM modules and creates a test-coverage report."
+        group = jsCheckTask
+
+        dependsOn(
+            auditNodePackages,
+            coverageJs,
+            testJs,
+        )
+    }
+
```
Comment (armiol) [inline buildSrc/src/main/kotlin/io/spine/internal/gradle/javascript/task/Check.kt:105]: Same here. 

Please review the changeset for this issue.

**Review by armiol — CHANGES_REQUESTED** [review]
24/40 reviewed.

Please see my comments so far.

**Review by armiol — COMMENTED** [review]
Please see another portion of the comments.

**Comment by armiol** [comment]
@yevhenii-nadtochii LGTM. Let's do this one more time, starting from `config`.

**Comment by yevhenii-nadtochii** [comment]
Closing this draft as it's approved and an associated PR into `config` repository is created.
See [Migrate `buildSrc/groovy/js/*` scripts to Kotlin](https://github.com/SpineEventEngine/config/pull/308)

**Pull request closed without merging**